server/src/main/java/org/elasticsearch/script/ScriptLanguagesInfo.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.script;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
+public class ScriptLanguagesInfo implements ToXContentObject, Writeable {
+    private static final ParseField TYPES_ALLOWED = new ParseField("types_allowed");
+    private static final ParseField LANGUAGE_CONTEXTS = new ParseField("language_contexts");
+    private static final ParseField LANGUAGE = new ParseField("language");
+    private static final ParseField CONTEXTS = new ParseField("contexts");
+
+    public final Set<String> typesAllowed;
+    public final Map<String,Set<String>> languageContexts;
+
+    public ScriptLanguagesInfo(Set<String> typesAllowed, Map<String,Set<String>> languageContexts) {
+        this.typesAllowed = typesAllowed != null ? Set.copyOf(typesAllowed): Collections.emptySet();
+        this.languageContexts = languageContexts != null ? Map.copyOf(languageContexts): Collections.emptyMap();
+    }
+
+    public ScriptLanguagesInfo(StreamInput in) throws IOException {
+        typesAllowed = readStringSet(in);
+        languageContexts = readStringMapSet(in);
```

Comment:
this can be `readMap(StreamInput::readString, in -> in.readSet(StreamInput::readString))`. Similar for the write side. None of the utility read/write methods here should be necessary.

Reply:
Done.